_posts/2020-10-13-from-aligned-to-unaligned-checkpoints-part-1.md

```diff
@@ -0,0 +1,117 @@
+
+
+## State in Streaming Applications
+
+Simply put, State is the information that you need to remember across events. Even the most trivial streaming applications are typically stateful because of their need to “remember” the exact position they are processing data from, in the form of a Kafka Partition Offset or a File Offset.
+In addition, many applications hold state internally as a way to support their internal operations, such as windows, aggregations, joins, or state machines.
+
+For the remainder of this article, we use the following example showing a streaming application consisting of four operators, each one holding some state
+
+<span>
+	<center>
+	<img vspace="8" style="width:75%" src="{{site.baseurl}}/img/blog/2020-10-13-from-aligned-to-unaligned-checkpoints-part-1/from-aligned-to-unaligned-checkpoints-part-1-imge-1.png" />
+	</center>
+</span>
+
+## State Persistence through Checkpoints
+
+Streaming applications are long-lived. They inevitably experience hardware and software failures but should, ideally, look from the outside as if no failure ever happened. Since applications are long-lived — and can potentially accumulate very large state —, recomputing partial results after failures can take quite some time and thus a way to persist and recover the (potentially very large) application state is necessary.
+
+Flink relies on its state checkpointing and recovery mechanism to implement such behavior, as shown in the figure below. Periodic checkpoints store a snapshot of the application’s state on some Checkpoint Storage (commonly an Object Store or Distributed File System, like S3, HDFS, GCS, Azure Blob Storage, etc.). When a failure is detected, the affected parts of the application are reset to the state of the latest checkpoint (either by a local reset or by loading the state from the checkpoint storage).
+
+<span>
+	<center>
+	<img vspace="8" style="width:100%" src="{{site.baseurl}}/img/blog/2020-10-13-from-aligned-to-unaligned-checkpoints-part-1/from-aligned-to-unaligned-checkpoints-part-1-imge-2.png" />
+	</center>
+</span>
+
+Flink’s checkpoint-based approach differs from the approach taken by other stream processing systems, for example keeping state in a distributed database, or writing state changes to a log. The checkpoint-based approach has some nice properties, described below, which make it a great option for Apache Flink.
+
+- **Checkpointing has very simple external dependencies**: An Object Storage or Distributed FileSystem are probably the most available and easiest-to-administer services. Because they are available on all public cloud providers and among the first systems to provide on-premises, Flink becomes a well suited system for a “Cloud Native” stack. Moreover, these storage systems are cheaper by an order of magnitude (GB/month) when compared to distributed databases, key/value stores, or event brokers.
```

Comment:
```suggestion
- **Checkpointing has very simple external dependencies**: An _Object Storage_ or a _Distributed FileSystem_ are probably the most available and easiest-to-administer services. Because these are available on all public cloud providers and among the first systems to provide on-premises, Flink becomes well-suited for a _cloud-native_ stack. In addition, these storage systems are cheaper by an order of magnitude (GB/month) when compared to distributed databases, key/value stores, or event brokers.
```